CBMC: Use instrumented malloc/free for MLD_ALLOC/MLD_FREE

hanno-becker opened the pull request:
* Port of: https://github.com/pq-code-package/mlkem-native/pull/1405

The default implementation of MLD_ALLOC and MLD_FREE uses stack
allocation, which the compiler can assume not to fail. This means
that CBMC does not exercise the cleanup paths which handle fallible
dynamic allocation -- a significant proof gap.

This commit changes the configuration uses for the CBMC proofs
to use the (instrumented) calls to malloc/free for MLD_ALLOC/MLD_FREE.

Importantly, we set --malloc-may-fail to model allocation as fallible,
and --malloc-fail-null to return NULL in case of allocation failure.

hanno-becker committed:
CBMC: Use instrumented malloc/free for MLD_ALLOC/MLD_FREE

The default implementation of MLD_ALLOC and MLD_FREE uses stack
allocation, which the compiler can assume not to fail. This means
that CBMC does not exercise the cleanup paths which handle fallible
dynamic allocation -- a significant proof gap.

This commit changes the configuration uses for the CBMC proofs
to use the (instrumented) calls to malloc/free for MLD_ALLOC/MLD_FREE.

Importantly, we set --malloc-may-fail to model allocation as fallible,
and --malloc-fail-null to return NULL in case of allocation failure.

Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

mkannwischer reviewed (commented):
Thanks @hanno-becker. The changes itself look good to me, but CBMC fails due to timeouts. Marking as draft until we can resolve that.

rod-chapman committed:
Add gcc to the NIX shell on macOS to support CBMC.

Leaves CC set as "clang" for compilation and testing
on macOS.

Always use gcc for pre-processing with goto-cc to get
consistent C header files on all platforms, including
macOS/Darwin.

Ported from mlkem-native (pq-code-package/mlkem-native#1780,
commit b828ccdb28db111695837efdbb3a85338b5ef112).

Signed-off-by: Rod Chapman <rodchap@amazon.com>
Signed-off-by: Hanno Becker <beckphan@amazon.co.uk>

mkannwischer reviewed (approved):
Thanks @hanno-becker. LGTM
